README.md
[![](https://ptk.dev/img/bot/social-manager-tools.png)](https://socialmanager.tools)

# 🤖 Social Manager Tools - Desktop App

[![](https://img.shields.io/badge/version-v0.6.10-lightgrey.svg)](https://github.com/social-manager-tools/socialmanagertools-gui/releases) [![](https://img.shields.io/badge/license-MIT-brightgreen.svg)](#) [![](https://img.shields.io/badge/ES-9-F7DF1E.svg)](https://wikipedia.org/wiki/ECMAScript) [![](https://img.shields.io/badge/powered%20by-puppeteer-46aef7.svg)](https://github.com/GoogleChrome/puppeteer) [![](https://snyk.io/test/github/social-manager-tools/socialmanagertools-gui/badge.svg)](https://snyk.io/test/github/social-manager-tools/socialmanagertools-gui)

> This Desktop Application helps you increase the engagement level of your Social Network profiles through different social algorithms: increase the likes on your photos and followers!

> ⛔ **DISCLAIMER**: This is an **unofficial** library and offers no warranty! The developers and contributors of the project do not assume any responsibility in case of ban of your account. Use bots does not comply with the social networks terms of the service: use this software at your own risk. A "bot" is legal software, but the use of a bot continuously violates the terms of use of Instagram, Twitter and other social and you risk a: __soft ban__ (such as limited actions or follow-up) or you risk __ban__ with suspension for a few days (or permanent). All trademarks and logos belong to their respective owners.

## 🎁 Support: Donate
> This project is **free**, **open source** and I try to provide excellent **free support**. Why donate? I work on this project several hours a week or in my spare time and try to keep it up to date and working. I do not intend to charge the basic features, but this cannot be done without your financial support, even small. There are professional bots with closed source on the Internet that save the password on the cloud for €14/month (for single social): we want to protect your password and offer you a better product than theirs. A lower donation would allow continuous development, ever better quality and the realization of this dream. **THANK YOU!**

[![](https://img.shields.io/badge/donate-paypal-005EA6.svg)](https://www.paypal.me/ptkdev) [![](https://img.shields.io/badge/donate-patreon-F87668.svg)](https://www.patreon.com/ptkdev) [![](https://img.shields.io/badge/buy%20me-coffee-4B788C.svg)](https://ko-fi.com/ptkdev)

![](https://img.shields.io/badge/bitcoin-35jQmZCy4nsxoMM3QPFrnZePDVhdKaHMRH-E38B29.svg) ![](https://img.shields.io/badge/ethereum-0x8b8171661bEb032828e82baBb0B5B98Ba8fBEBFc-4E8EE9.svg)

## 📎 Menu
- 💡 [Features](#-features)
- 💾 Installation:
  - 👨‍💻 [Desktop App](http://docs.socialmanager.tools/gui/installation/README.md) (Windows, Linux, Mac)
  - 🔨 [Cli](http://docs.socialmanager.tools/igbot/installation/source/README.md) (CLI / NPM Package)
  - 🦀 [Raspberry PI](http://docs.socialmanager.tools/igbot/installation/raspberry/README.md) (Raspbian 9)
  - 🐧 [Linux Server](http://docs.socialmanager.tools/igbot/installation/linux/README.md) (Debian 9 Server)
  - 🐳 [Docker](http://docs.socialmanager.tools/igbot/installation/docker/README.md)
- 🔑 Configs:
  - 💿 [List of Mode](http://docs.socialmanager.tools/igbot/configs/modes/README.md)
  - 📀 [How set Multi-account](http://docs.socialmanager.tools/igbot/configs/multiaccount/README.md)
- 📚 [API Documentation](http://docs.socialmanager.tools/igbot/api/README.md) (How create your personal bot or mode)
- 💻 [Developers Guidelines](http://docs.socialmanager.tools/developers/guidesline/README.md) (If you want contributing)
- 🐛 [Known Bugs](https://github.com/social-manager-tools/socialmanagertools-gui/issues?q=is%3Aopen+is%3Aissue+label%3Abug)
- 🍻 Community:
  - 🐦 [Telegram group](http://telegram.ptkdev.io)
  - 🐔 [Discord](http://discord.ptkdev.io)
  - 🐓 [Slack](http://slack.ptkdev.io)
  - 🕊 [Medium](http://blog.ptkdev.io)
  - 🐤 [Twitter](http://twitter.ptkdev.io)

## 🔖 Screenshot
[![](https://ptk.dev/img/bot/social-manager-tools.gif)](https://instagram.bot.ptkdev.io)

## 💡 Features
* [✓] Easy to use
* [✓] GUI for [socialmanagertools-igbot](https://github.com/social-manager-tools/socialmanagertools-igbot) library.
* [✓] GUI for [socialmanagertools-twbot](https://github.com/social-manager-tools/socialmanagertools-twbot) library.
* [✘] GUI for [socialmanagertools-fbbot](https://github.com/social-manager-tools/socialmanagertools-fbbot) library.
* [✘] GUI for [socialmanagertools-tgbot](https://github.com/social-manager-tools/socialmanagertools-tgbot) library.

## 👨‍💻 Install app
1. Download [stable version](https://socialmanager.tools) or see other [packages](https://github.com/social-manager-tools/socialmanagertools-gui/releases)
2. Run installation.

## 🔧 Developer Setup
1. Download [stable version](https://github.com/social-manager-tools/socialmanagertools-gui/releases) and extract it.
2. Download [Node.js](https://nodejs.org/it/) >= 7.6 and install it.
3. Run `npm install electron electron-builder -g --unsafe-perm=true --allow-root`
4. Run `npm install` in `social-manager-tools-gui` folder.
5. If you received error with sqllite3 problem, run `npm run build-fix`
6. Start the ui bot via `npm run start` or build via `npm run dist`
7. If work add star :star: at this project :heart:
8. If you want help me: **[donate on paypal](http://paypal.ptkdev.io)** or become a **[backer on patreon](http://patreon.ptkdev.io)**.

## 👑 Sponsors
Support this project by becoming a sponsor. 🙏 Become a sponsor on [patreon](https://www.patreon.com/join/ptkdev) or become top3 sponsor on [ko-fi](https://ko-fi.com/ptkdev). Your logo will show up here with a link to your website.

[![](https://api.ptkdev.io/backers/sponsor1.png)](https://api.ptkdev.io/backers/sponsor1.html) [![](https://api.ptkdev.io/backers/sponsor2.png)](https://api.ptkdev.io/backers/sponsor2.html) [![](https://api.ptkdev.io/backers/sponsor-kofi1.png)](https://api.ptkdev.io/backers/sponsor-kofi1.html) [![](https://api.ptkdev.io/backers/sponsor-kofi2.png)](https://api.ptkdev.io/backers/sponsor-kofi2.html) [![](https://api.ptkdev.io/backers/sponsor-kofi3.png)](https://api.ptkdev.io/backers/sponsor-kofi3.html) [![](https://api.ptkdev.io/backers/sponsor3.png)](https://api.ptkdev.io/backers/sponsor3.html) [![](https://api.ptkdev.io/backers/sponsor4.png)](https://api.ptkdev.io/backers/sponsor4.html) [![](https://api.ptkdev.io/backers/sponsor5.png)](https://api.ptkdev.io/backers/sponsor5.html) [![](https://api.ptkdev.io/backers/sponsor6.png)](https://api.ptkdev.io/backers/sponsor6.html) [![](https://api.ptkdev.io/backers/sponsor7.png)](https://api.ptkdev.io/backers/sponsor7.html) [![](https://api.ptkdev.io/backers/sponsor8.png)](https://api.ptkdev.io/backers/sponsor8.html) [![](https://api.ptkdev.io/backers/sponsor9.png)](https://api.ptkdev.io/backers/sponsor9.html) [![](https://api.ptkdev.io/backers/sponsor10.png)](https://api.ptkdev.io/backers/sponsor10.html) [![](https://api.ptkdev.io/backers/sponsor11.png)](https://api.ptkdev.io/backers/sponsor11.html) [![](https://api.ptkdev.io/backers/sponsor12.png)](https://api.ptkdev.io/backers/sponsor12.html) [![](https://api.ptkdev.io/backers/sponsor13.png)](https://api.ptkdev.io/backers/sponsor13.html) [![](https://api.ptkdev.io/backers/sponsor14.png)](https://api.ptkdev.io/backers/sponsor14.html) [![](https://api.ptkdev.io/backers/sponsor15.png)](https://api.ptkdev.io/backers/sponsor15.html)

## 🦄 Backers
Thank you to all our backers! 🙏 Become a backer on [patreon](https://www.patreon.com/join/ptkdev).

[![](https://api.ptkdev.io/backers/backer1.png)](https://api.ptkdev.io/backers/backer1.html) [![](https://api.ptkdev.io/backers/backer2.png)](https://api.ptkdev.io/backers/backer2.html) [![](https://api.ptkdev.io/backers/backer3.png)](https://api.ptkdev.io/backers/backer3.html) [![](https://api.ptkdev.io/backers/backer4.png)](https://api.ptkdev.io/backers/backer4.html) [![](https://api.ptkdev.io/backers/backer5.png)](https://api.ptkdev.io/backers/backer5.html) [![](https://api.ptkdev.io/backers/backer6.png)](https://api.ptkdev.io/backers/backer6.html) [![](https://api.ptkdev.io/backers/backer7.png)](https://api.ptkdev.io/backers/backer7.html) [![](https://api.ptkdev.io/backers/backer8.png)](https://api.ptkdev.io/backers/backer8.html) [![](https://api.ptkdev.io/backers/backer9.png)](https://api.ptkdev.io/backers/backer9.html) [![](https://api.ptkdev.io/backers/backer10.png)](https://api.ptkdev.io/backers/backer10.html) [![](https://api.ptkdev.io/backers/backer11.png)](https://api.ptkdev.io/backers/backer11.html) [![](https://api.ptkdev.io/backers/backer12.png)](https://api.ptkdev.io/backers/backer12.html) [![](https://api.ptkdev.io/backers/backer13.png)](https://api.ptkdev.io/backers/backer13.html) [![](https://api.ptkdev.io/backers/backer14.png)](https://api.ptkdev.io/backers/backer14.html) [![](https://api.ptkdev.io/backers/backer15.png)](https://api.ptkdev.io/backers/backer15.html)

## ❤️ Contributing
I 💟 contributions! I will happily accept your pull request! Translations, grammatical corrections (GrammarNazi you are welcome! Yes my English is bad, sorry), new modes, best css selectors, fix and new feature! Read [Developers Guidelines](http://docs.socialmanager.tools/developers/guidesline/README.md) for best practices. Do not be afraid, if the code is not perfect we will work together 👯 and remember to insert your name in `.all-contributorsrc` and `package.json` file.

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://ptk.dev"><img src="https://avatars1.githubusercontent.com/u/442844?v=4" width="100px;" alt=""/><br /><sub><b>Patryk Rzucidło</b></sub></a><br /><a href="https://github.com/ptkdev/socialmanagertools-gui/commits?author=ptkdev" title="Code">💻</a></td>
    <td align="center"><a href="https://twitter.com/maxfer75"><img src="https://raw.githubusercontent.com/social-manager-tools/socialmanagertools-igbot/master/.github/assets/avatar/maxfer75.jpg" width="100px;" alt=""/><br /><sub><b>MaXfer </b></sub></a><br /><a href="#design-maxfer75" title="Design">🎨</a></td>
    <td align="center"><a href="https://twitter.com/tanik72"><img src="https://raw.githubusercontent.com/social-manager-tools/socialmanagertools-igbot/master/.github/assets/avatar/tanik72.jpg" width="100px;" alt=""/><br /><sub><b>Tanik</b></sub></a><br /><a href="#design-TaniK72" title="Design">🎨</a></td>
    <td align="center"><a href="https://github.com/Gaudo"><img src="https://avatars1.githubusercontent.com/u/708636?v=4" width="100px;" alt=""/><br /><sub><b>Francesco Gaudenzi</b></sub></a><br /><a href="https://github.com/ptkdev/socialmanagertools-gui/commits?author=Gaudo" title="Code">💻</a></td>
    <td align="center"><a href="https://github.com/NicolaCorvo90"><img src="https://avatars1.githubusercontent.com/u/17699027?v=4" width="100px;" alt=""/><br /><sub><b>Nicola Corvo</b></sub></a><br /><a href="https://github.com/ptkdev/socialmanagertools-gui/commits?author=NicolaCorvo90" title="Code">💻</a> <a href="https://github.com/ptkdev/socialmanagertools-gui/issues?q=author%3ANicolaCorvo90" title="Bug reports">🐛</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

> 💰 In the future, if the donations allow it, I would like to share some of the success with those who helped me the most. For me open source is share of code, share development knowledges and share donations!

## 📲 Tools

CHANGELOG.md
# v0.6.10 (30 March 2020)
> Update: [socialmanagertools-igbot v0.9.23](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Fix: fdfmode_classic


# v0.6.9 (25 March 2020)
> Update: [socialmanagertools-igbot v0.9.21](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Upgrade: puppeteer 2.1.1 (chrome v80 required)
* Fix: likemode_realistic
* Fix: likemode_classic
* Fix: likemode_superlike
* Fix: commentmode_classic
* Fix: fdfmode_classic


# v0.6.8 (02 December 2019)
* Limits update: after august 2019 instagram decrease action/daily from 1000 to 300. If bot excede limits? Instagramm will ask you to change the password.
* Fix: app freeze after start with fdf mode.


# v0.6.7 (28 September 2019)
> Update: [socialmanagertools-igbot v0.9.19](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Upgrade: puppeteer 1.19.0 (chrome v77 required)
* Upgrade: electron to v6
* Package: different deb file for Ubuntu/Debian
* Package: windows 10 build now has installer
* Fix: UI issues (Show button on try icon)


# v0.6.6 (10 July 2019)
> Update: [socialmanagertools-igbot v0.9.18](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Fix: UI issues.


# v0.6.5 (15 May 2019)
> Update: [socialmanagertools-igbot v0.9.17](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Perfomance: better, i tried fix lag of ui
* Upgrade: electron 5
* Upgrade: puppeteer 1.13.0
* Fix: password with double quotes #54
* Fix: min like/day #58
* Fix: waiting for selector failed: timeout 5000ms exceeded (mac only)
* Fix: Can't launch Chrome if SMT is running #38 (mac only - beta fix)
* Fix: minimize (linux gnome only - beta fix)
* Fix: if try open link on smt, popup and drawer go to close
* Fix: mac chrome open issue
* New: windows 32bit build
* New: community! Telegram and Discord
* New: patreon benefits!
* New: contributors to README.md


# v0.6.4 (01 April 2019)
> Update: [socialmanagertools-igbot v0.9.14](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Fix: `waiting for selector "article div a:nth-child(1)" failed: timeout 30000ms exceeded` happen when photo don't exist or are removed, now skip faster (3sec)
* Fix: docker curl package missing
* Fix: docker permissions
* Fix: README.md misprint
* Downgrade: puppeteer v1.11.0


# v0.6.3 (28 March 2019)
> Update: [socialmanagertools-igbot v0.9.11](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* New: `instagrambotlib` and `instagram-bot.js` now is `socialmanagertools-igbot`
* New: npm package org: `npm install @social-manager-tools/igbot`
* Deprecated: rolling release version
* Fix: Login
* Fix: fdfmode_classic
* Upgrade: docker images
* Upgrade: puppeteer v1.12.2


# v0.6.2 (28 November 2018)
> Update: [socialmanagertools-igbot v0.9.10](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Developers: [Guidelines](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/nightly/DEV_GUIDELINES.md) for contributing at project.
* Test: add mocha and chai
* Fix: fdfmode_classic loop bug
* Upgrade: puppeteer v1.7.0


# v0.6.1 (16 October 2018)
> Fix: [socialmanager-gui](https://github.com/social-manager-tools/socialmanagertools-gui/), changelog:
* Fix: Wrong instructions on where to find Post you've liked #41
> Update: [socialmanagertools-igbot v0.9.9](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Fix: fdfmode_classic #42 (again again...)
* Fix: 2FA (Location)
* Fix: Close button (ui)
* Fix: lose connection bot stop, now retry work.

# v0.6.0 (14 October 2018)
> Fix: [socialmanager-gui](https://github.com/social-manager-tools/socialmanagertools-gui/), changelog:
* Fix: bot remove space in comment_mode (socialmanagertools-igbot)
> Update: [socialmanagertools-igbot v0.9.8](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Fix: fdfmode_classic #42 (again)
* Fix: 2FA doesn't work #48 (after 3 months... thanks instagram!)
* Update: default value of `chrome_headless` is `true`
* Update: all library of package.json


# v0.5.6 (05 October 2018)
> Update: [socialmanagertools-igbot v0.9.6](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Fix: login random don't work
* Fix: fdfmode_classic #42


# v0.5.5 (02 October 2018)
> Update: [socialmanagertools-igbot v0.9.5](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md), changelog:
* Fix: likemode_classic new instagram selectors
* Fix: likemode_realistic new instagram selectors
* Fix: likemode_superlike new instagram selectors
* Fix: likemode_competitor_users new instagram selectors
* Fix: comment_mode new instagram selectors
* Fix: fdfmode_classic new instagram selectors
* API: Fix stop() again, now really work.
> Update: [socialmanagertools-twbot v0.2.5](https://github.com/social-manager-tools/socialmanagertools-twbot/blob/master/CHANGELOG.md), changelog:
* Fix: Startup error


# v0.5.4 (26 September 2018)
* Update: [socialmanagertools-igbot v0.9.4](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md)
* Update: [socialmanagertools-twbot v0.2.4](https://github.com/social-manager-tools/socialmanagertools-twbot/blob/master/CHANGELOG.md)


# v0.5.3 (22 September 2018)
* Update: [socialmanagertools-igbot v0.9.3](https://github.com/social-manager-tools/socialmanagertools-igbot/blob/master/CHANGELOG.md)
* Fix: various misprint and minor bug

src/js/smt-instagram.js


function instagrambot_start(json, id) {
    const main = require("electron").remote.require("./main");
    main.instagrambot_start(json, id);
}

function instagrambot_stop(id) {
    delete list_actived_bot[id];
    const main = require("electron").remote.require("./main");
    main.instagrambot_stop(id);
    app.dialog.create({ title: "Status", text: "Stopped...<br />" + donatehtml, buttons: [{ text: "OK" }] }).open();
}

function instagram_get_user_form() {
    let tokens = {};
    tokens.instagram_username = $("#instagram_username").val();
    tokens.instagram_username = tokens.instagram_username.replace(/@/g, "");
    tokens.instagram_password = $("#instagram_password").val();
    tokens.instagram_password = tokens.instagram_password.replace(/"/g, "\\\"");
    tokens.instagram_hashtag = $("#instagram_hashtag").val();
    tokens.bot_mode = $("#bot_mode").val();
    tokens.executable_path = $("#executable_path").val();
    tokens.instagram_hashtag = tokens.instagram_hashtag.replace(/ /g, "");
    tokens.instagram_hashtag = tokens.instagram_hashtag.replace(/#/g, "");
    tokens.instagram_hashtag = tokens.instagram_hashtag.replace(/,/g, "\",\"");

    tokens.bot_likeday_min = parseInt($("#bot_likeday_max").val() - 100);
    tokens.bot_likeday_max = $("#bot_likeday_max").val();
    if(tokens.bot_likeday_min <= 100){
        tokens.bot_likeday_min = 120;
        tokens.bot_likeday_max = 150;
    }

    tokens.bot_followday = $("#bot_followday").val();
    tokens.bot_followrotate = $("#bot_followrotate").val();
    tokens.bot_userwhitelist = $("#bot_userwhitelist").val();
    tokens.bot_userwhitelist = tokens.bot_userwhitelist.replace(/ /g, "");
    tokens.bot_userwhitelist = tokens.bot_userwhitelist.replace(/#/g, "");
    tokens.bot_userwhitelist = tokens.bot_userwhitelist.replace(/,/g, "\",\"");

    if (tokens.bot_mode == "comment_mode") {
        tokens.bot_likeday_min = parseInt($("#bot_commentsday").val() - 50);
        tokens.bot_likeday_max = $("#bot_commentsday").val();
    }

    tokens.bot_comment_list = $("#bot_comment_list").val();
    tokens.bot_comment_list = tokens.bot_comment_list.replace(/,/g, "\",\"");

    tokens.bot_likemode_competitor_users = $("#bot_likemode_competitor_users").val();


    let min = 0;
    do {
        min++;
    } while ((60 / min * 24 * 11) > tokens.bot_likeday_min);
    tokens.bot_fastlike_max = min--;
    min = 0;
    do {
        min++;
    } while ((60 / min * 24 * 11) > tokens.bot_likeday_max);
    tokens.bot_fastlike_min = min--;
    tokens.bot_superlike_n = $("#bot_superlike_n").val();

    if ($("#chrome_headless").val() != "enabled" && $("#chrome_headless").val() != "disabled") {
        tokens.chrome_headless = "disabled";
    } else {
        tokens.chrome_headless = $("#chrome_headless").val();
    }

    if (process.platform === "win32") {
        tokens.config_path = remoteapp.getPath("userData");
        tokens.config_path = tokens.config_path.replace(/\\/g, "/");
    } else {
        tokens.config_path = remoteapp.getPath("userData");
    }

    return tokens;
}

function instagram_load_config() {
    if ($("#executable_path").val() == "") {
        if (process.platform === "linux") {
            $("#executable_path").val("/usr/bin/google-chrome");
        } else if (process.platform === "win32") {
            $("#executable_path").val("C:/Program Files (x86)/Google/Chrome/Application/chrome.exe");
        } else if (process.platform === "darwin") {
            $("#executable_path").val("/Applications/Google Chrome.app/Contents/MacOS/Google Chrome");
        }
    }

    $("#bot_mode").change();

    fs.exists(remoteapp.getPath("userData") + "/config_" + $("#instagram_username").val() + ".json", function(exists) {
        if (exists) {
            let data = fs.readFileSync(remoteapp.getPath("userData") + "/config_" + $("#instagram_username").val() + ".json", "utf8");
            let config = JSON.parse(data.toString());

            $("#instagram_password").val(config.instagram_password);
            $("#bot_mode").val(config.bot_mode);
            $("#bot_likeday_max").val(config.bot_likeday_max);
            $("#bot_superlike_n").val(config.bot_superlike_n);
            $("#executable_path").val(config.executable_path);
            $("#chrome_headless").val(config.chrome_headless);
            $("#instagram_hashtag").val(config.instagram_hashtag.join());

            $("#bot_commentsday").val(config.bot_likeday_max);
            $("#bot_comment_list").val(config.comment_mode.comments.source.join());
            $("#bot_followday").val(config.bot_followday);
            $("#bot_followrotate").val(config.bot_followrotate);
            $("#bot_userwhitelist").val(config.bot_userwhitelist.join());
            $("#bot_likemode_competitor_users").val(config.likemode_competitor_users.account);

            instagram_check_bot_mode();
        }
    });

}

function instagram_check_form() {
    let check_err = 0;

    if ($("#instagram_username").val() == "" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Username is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#instagram_password").val() == "" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Password is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#instagram_hashtag").val() == "" && $("#bot_mode").val() != "likemode_competitor_users" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Hashtag list is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#bot_likeday_max").val() == "" && $("#bot_mode").val() != "fdfmode_classic" && $("#bot_mode").val() != "comment_mode" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Max like/day is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#bot_mode").val() == "likemode_superlike" && $("#bot_superlike_n").val() == "" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Max Like photo/user is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#bot_commentsday").val() == "" && $("#bot_mode").val() === "comment_mode" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Max Comments/day is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#bot_comment_list").val() == "" && $("#bot_mode").val() === "comment_mode" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Comments list is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#bot_followday").val() == "" && $("#bot_mode").val() === "fdfmode_classic" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Max follow-defollow/day is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#bot_followrotate").val() == "" && $("#bot_mode").val() === "fdfmode_classic" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Max follow rotate is empty", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if ($("#bot_likemode_competitor_users").val() == "" && $("#bot_mode").val() === "likemode_competitor_users" && check_err == 0) {
        app.dialog.create({ title: "Warning", text: "Insert competitor username", buttons: [{ text: "OK" }] }).open();
        check_err++;
    } else if (check_err == 0) {
        instagram_check_max_like();
    }

    return check_err;
}

function instagram_save_config(bot) {
    let check_err = instagram_check_form();

    if (list_actived_bot[$("#instagram_username").val()+"_"+$("#bot_mode").val()] == true) {
        app.dialog.create({ title: "Status", text: "@" + $("#instagram_username").val() + " bot is running in this mode... Stop this mode and start again if you need change this configuration", buttons: [{ text: "OK" }] }).open();
        return 1;
    } else if (bot === "instagram" && check_err <= 0) {
        let tokens = instagram_get_user_form();
        let json = jtr.replace(tokens, require("../config.json"));
        // clean old logs
        fs.writeFile(remoteapp.getPath("userData") + "/" + $("#instagram_username").val() + "_" + $("#bot_mode").val() + ".log", "", function(err) {
            if (err) {
                return console.log(err);
            }
        });

        fs.writeFile(remoteapp.getPath("userData") + "/config_" + $("#instagram_username").val() + ".json", JSON.stringify(json), function(err) {
            if (err) {
                return console.log(err);
            }

            fs.exists($("#executable_path").val(), function(exists) {
                if (exists) {
                    app.dialog.create({ title: "Status", text: "Bot started...<br />" + donatehtml, buttons: [{ text: "OK" }] }).open();
                    let id = tokens.instagram_username+"_"+tokens.bot_mode;
                    list_actived_bot[id] = true;
                    instagrambot_start(json, id);
                } else {
                    app.dialog.create({ title: "Warning", text: "Google Chrome path doesn't exist, please install google chrome or chromium and retry", buttons: [{ text: "OK" }] }).open();
                }
            });
        });
    }
}

function instagram_save_2fa() {

    fs.writeFile(remoteapp.getPath("userData") + "/" + $("#instagram_username").val() + "_pin.txt", $("#instagram_2fapin").val(), function(err) {
        if (err) {
            return console.log(err);
        }
    });
}

function instagram_open_logs() {
    clearInterval(logs_interval);
    $("#logs").val("[INFO] Loading...");
    logs_interval = setInterval(function() {
        let logs = fs.readFileSync(remoteapp.getPath("userData") + "/" + $("#instagram_username").val() + "_" + $("#bot_mode").val() + ".log", "utf8");
        $("#logs").val(logs);
    }, 1000);
}

function instagram_check_bot_mode() {
    if ($("#bot_mode").val() == "likemode_classic") {
        $(".bot_mode_desc").html("Bot go to random hashtag from list, like 1 photo and stop X minutes in loop");
        $(".likemode_all").hide();
        $("#bot_superlike_n").val(3);
        $("#bot_followday").val(250);
        $("#bot_followrotate").val(30);
        $("#bot_commentsday").val(300);
        $(".likemode_classic").show();
    } else if ($("#bot_mode").val() == "likemode_realistic") {
        $(".bot_mode_desc").html("Bot go to random hashtag from list, like 10-12 photo fast and stop X minutes in loop");
        $(".likemode_all").hide();
        $("#bot_superlike_n").val(3);
        $("#bot_followday").val(250);
        $("#bot_followrotate").val(30);
        $("#bot_commentsday").val(300);
        $(".likemode_realistic").show();
    } else if ($("#bot_mode").val() == "likemode_superlike") {
        $(".bot_mode_desc").html("Bot go to random hashtag from list, go to author profile, like 3 (configurable) random photos, return to hashtag list for 10-11 times and stop X minutes in loop");
        $(".likemode_all").hide();
        $("#bot_followday").val(250);
        $("#bot_followrotate").val(30);
        $("#bot_commentsday").val(250);
        $(".likemode_superlike").show();
    } else if ($("#bot_mode").val() == "likemode_competitor_users") {
        $(".bot_mode_desc").html("(Bot go to competitor account, like photos of his followers (10-12 photo fast and stop X minutes in loop))");
        $(".likemode_all").hide();
        $("#bot_superlike_n").val(3);
        $("#bot_followday").val(250);
        $("#bot_followrotate").val(30);
        $("#bot_commentsday").val(250);
        $(".likemode_competitor_users").show();
    } else if ($("#bot_mode").val() == "comment_mode") {
        $(".bot_mode_desc").html("Bot go to random hashtag from list, leave random comments from comments-list");
        $(".likemode_all").hide();
        $("#bot_superlike_n").val(3);
        $("#bot_followday").val(250);
        $("#bot_followrotate").val(30);
        $(".comment_mode").show();
    } else if ($("#bot_mode").val() == "fdfmode_classic") {
        $(".bot_mode_desc").html("Bot go to random hashtag from list, follow 30 users fast at 31 defollow first followed (number 1), follow 32, defollow number 2, in loop.");
        $(".likemode_all").hide();
        $("#bot_superlike_n").val(3);
        $("#bot_likeday_max").val($("#bot_followday").val());
        $(".fdfmode_classic").show();
    }
}

function instagram_check_hashtag() {
    return 1;
}

function instagram_check_superlike_n() {
    if ($("#bot_mode").val() == "likemode_superlike" && (parseInt($("#bot_superlike_n").val() * $("#bot_likeday_max").val()) >= 300)) {
        app.dialog.create({ title: "Warning", text: $("#bot_superlike_n").val() + "x" + $("#bot_likeday_max").val() + " = is more than 300 like/day. After august 2019 instagram decrease action/daily from 1000 to 300.", buttons: [{ text: "OK" }] }).open();

        return 1;
    }

    return 0;
}

function instagram_check_max_like() {
    if ($("#bot_mode").val() == "likemode_superlike" && (parseInt($("#bot_superlike_n").val() * $("#bot_likeday_max").val()) >= 300)) {
        app.dialog.create({ title: "Warning", text: $("#bot_superlike_n").val() + "x" + $("#bot_likeday_max").val() + " = is more than 300 like/day. After august 2019 instagram decrease action/daily from 1000 to 300.", buttons: [{ text: "OK" }] }).open();

        return 1;

    } else if ($("#bot_likeday_max").val() != "" && parseInt($("#bot_likeday_max").val()) > 300) {
        app.dialog.create({ title: "Warning", text: "Instagram ask you to change password if set likes, comments or follow/defollow more than 300/daily. After august 2019 instagram decrease action/daily from 1000 to 300.", buttons: [{ text: "OK" }] }).open();

        return 1;
    }

    return 0;
}


function instagram_check_likemode_competitor_users() {
    if ($("#bot_likemode_competitor_users").val() == "") {
        app.dialog.create({ title: "Warning", text: "Insert competitor username, restore default value", buttons: [{ text: "OK" }] }).open();

        $("#bot_likemode_competitor_users").val("ptkdev");

        return 1;
    }

    return 0;
}

function instagram_check_followrotate() {
    if ($("#bot_followrotate").val() > 3000) {
        app.dialog.create({ title: "Warning", text: "Value follow rotate is overcapacity, restore default value", buttons: [{ text: "OK" }] }).open();

        $("#bot_followrotate").val(30);

        return 1;
    }

    return 0;
}

function instagram_check_followday() {
    if ($("#bot_followday").val() > 300) {
        app.dialog.create({ title: "Warning", text: "Instagram ask you to change password if set likes more than 300follow/daily. After august 2019 instagram decrease action/daily from 1000 to 300.", buttons: [{ text: "OK" }] }).open();

        return 1;
    }

    return 0;
}

function instagram_check_commentsday() {
    if ($("#bot_commentsday").val() > 1000) {
        app.dialog.create({ title: "Warning", text: "Value comments/daily is overcapacity. After august 2019 instagram decrease action/daily from 1000 to 300.", buttons: [{ text: "OK" }] }).open();

        return 1;
    }

    return 0;
}

function instagram_check_userwhitelist() {
    return 1;
}


function instagram_check_comment_mode() {
    return 1;
}
